table/src/main/java/com/bazaarvoice/emodb/table/db/astyanax/AstyanaxTableDAO.java

```diff
@@ -1559,4 +1640,214 @@ protected static String getMillisecondPrecisionZonedDateTime(ZonedDateTime dateT
+                            .stream()
+                            .map(DataCenter::getName)
+                            .anyMatch(dataCenter -> dataCenter.equals(tableEventDatacenter.getDataCenter())))
+                    .map(storage ->
+                            tableEventDatacenter.newTableEvent(table, new TableEvent(action, storage.getUuidString()), _clock.instant())
+                    )
+                    .map(delta -> {
+                        Audit audit = new AuditBuilder()
+                                .setComment(String.format("Adding events for table %s", table))
+                                .setLocalHost()
+                                .build();
+                        return new Update(_systemTableEventRegistry, tableEventDatacenter.getDataCenter(), TimeUUIDs.newUUID(), delta, audit, WriteConsistency.GLOBAL);
+                    })
+                    .forEach(updates::add);
+        });
+
+        _backingStore.updateAll(updates);
+    }
+
+    private void addMoveTableEvent(TableJson json, Storage srcStorage, String destUuid, String destPlacement) {
+        Iterator<Map<String, Object>> tableEventDatacenterIterator = _backingStore.scan(_systemTableEventRegistry, null, LimitCounter.max(), ReadConsistency.STRONG);
+
+        Collection<DataCenter> srcDatacenters = _placementFactory.getDataCenters(srcStorage.getPlacement());
+
+        Collection<DataCenter> destDatacenters = _placementFactory.getDataCenters(destPlacement);
+
+        Collection<String> dropDatacenters = srcDatacenters
+                .stream()
+                .filter(datacenter -> !destDatacenters.contains(datacenter))
+                .map(DataCenter::getName)
+                .collect(Collectors.toList());
+
+        Collection<String> promoteDatacenters = destDatacenters
+                .stream()
+                .filter(datacenter -> !srcDatacenters.contains(datacenter))
+                .map(DataCenter::getName)
+                .collect(Collectors.toList());
+
+        List<Update> updates = StreamSupport.stream(Spliterators.spliteratorUnknownSize(tableEventDatacenterIterator, 0), false)
+                .map(datacenter -> _objectMapper.convertValue(datacenter, TableEventDatacenter.class))
+                .flatMap(tableEventDatacenter -> {
+
+                    Delta delta = null;
+
+                    if (dropDatacenters.contains(tableEventDatacenter.getDataCenter())) {
+                        delta = tableEventDatacenter.newTableEvent(json.getTable(), new TableEvent(TableEvent.Action.DROP, srcStorage.getUuidString()), _clock.instant());
+                    } else if (promoteDatacenters.contains(tableEventDatacenter.getDataCenter())) {
+                        delta = tableEventDatacenter.newTableEvent(json.getTable(), new TableEvent(TableEvent.Action.PROMOTE, destUuid), _clock.instant());
+                    } else {
+                        return Stream.empty();
```

Comment:
This can indeed occur in our existing datacenters. Consider the case where move a table from ugc_global to catalog_global. There would be no table events necessary,

Reply:
alright! 👍